lib/puppet/indirector/json.rb

```diff
@@ -52,7 +52,7 @@ def load_json_from_file(file, key)
     json = nil
 
     begin
-      json = File.read(file)
+      json = File.read(file).force_encoding(Encoding::UTF_8)
```

Comment:
Forcing the encoding is only safe if we're positive the contents were originally written as UTF-8, but I'm not sure that's guaranteed. The [`save`](https://github.com/branan/puppet/blob/bug/stable/pup-5584-cached-catalog-encoding/lib/puppet/indirector/json.rb#L18) method calls replace_file, which opens a uniquefile, but doesn't specify an explicit encoding. It then calls `IO#print to_json` which as far as I can tell will transcode the string from the internal default encoding to the external default encoding, but according to http://ruby-doc.org/core-2.1.4/Encoding.html#class-Encoding-label-Internal+encoding, that only happens if the internal default encoding is non-nil. On *nix, I get:

```
$ ruby -e 'p Encoding.default_internal'
nil
$ ruby -e 'p Encoding.default_external'
#<Encoding:UTF-8>
```

On Windows, internal is nil and default external depends on the codepage:

```
C:\Program Files\Puppet Labs\Puppet\sys\ruby\bin> ruby -e 'p Encoding.default_internal'
nil
C:\Program Files\Puppet Labs\Puppet\sys\ruby\bin> ruby -e 'p Encoding.default_external'
#<Encoding:IBM437>
```
In other words, I *think* the `save` method will write the string to the file preserving whatever encoding the string is currently in, which may not be UTF-8. For example, if `to_json` returns a UTF-16LE string (for some reason) whose byte representation is [0xDE, 0x00, 0xAD, 0x00, 0xBE, 0x00, 0xEF, 0x00], then the file would contain those bytes. If we then read the file back in and force encoding to UTF-8, the string would only contain 0xDE, as the next byte would be interpreted as the NULL terminator.

We might need to modify the `save` method to ensure we're always writing a UTF-8 string, e.g.

```ruby
f.print to_json(request.instance).encode('UTF-8')
```

and then force_encoding when we read it back in.